source/source.go

```diff
@@ -30,20 +31,65 @@ import (
 
 type Source struct {
 	sdk.UnimplementedSource
-	config   Config
-	svc      *sqs.Client
-	queueURL string
+	config        Config
+	svc           *sqs.Client
+	queueURL      string
+	savedMessages *savedMessages
```

Comment:
My 2 cents - I'd change this type to `[]types.Message` and store the response of `s.svc.ReceiveMessage` in this field. In `receiveMessage` we can then just read from this slice and shrink it until it's 0. Once it's 0, we again do a request and store the response.